### PR TITLE
Fixes #3298 - Rework ServerContext PostStartup logic

### LINKED ISSUE #3298: Fixes for post-startup validation
Only need to run post-startup validation when there are post-startup tasks pending (e.g. replications).  

Also need to clean up handling when https is enabled.

Add always-enabled logging prior to/after post-startup tasks when they are run, to make it more transparent in the logs what's going on.

### PR DESCRIPTION
**Pull request by bbrks**
Fixes #3298 - Broken HTTPS startup

Reverts the changes made in #3254 and simplifies the wait logic.

### CODE REVIEW

**Comment by bbrks** [comment]
@adamcfraser  New `PostStartup` logic now delays replicator startup by 1 second to allow APIs to fire up. PostStartup is fired _after_ `AddDatabaseFromConfig` to allow for re-indexing.

```
2018-02-13T18:21:10.627Z Starting admin server on 127.0.0.1:4985
2018-02-13T18:21:10.630Z Starting server on :4984 ...
2018-02-13T18:21:11.632Z Replicate: Starting replication with parameters {ReplicationId:rep1 Source:http://127.0.0.1:4985 SourceDb:db1 Channels:[repchan1] Target:http://127.0.0.1:4985 TargetDb:db2 ChangesFeedLimit:100 Lifecycle:0 Disabled:false Async:true}
2018-02-13T18:21:11.633Z Replicate: Starting replication with parameters {ReplicationId:rep2 Source:http://127.0.0.1:4985 SourceDb:db1 Channels:[repchan2] Target:http://127.0.0.1:4985 TargetDb:db2 ChangesFeedLimit:100 Lifecycle:0 Disabled:false Async:true}
2018/02/13 18:21:11 Replicate: [rep1] stateFnPreStarted got event: {REPLICATION_START <nil>}
2018/02/13 18:21:11 Replicate: [rep1] stateFnPreStarted got START event: {REPLICATION_START <nil>}
2018/02/13 18:21:11 Replicate: [rep2] stateFnPreStarted got event: {REPLICATION_START <nil>}
2018/02/13 18:21:11 Replicate: [rep2] stateFnPreStarted got START event: {REPLICATION_START <nil>}
```

File: rest/server_context.go [hunk 1]
```diff
@@ -87,18 +87,24 @@ func NewServerContext(config *ServerConfig) *ServerContext {
 	return sc
 }
 
+// PostStartup runs anything that relies on SG being fully started (i.e. sgreplicate)
 func (sc *ServerContext) PostStartup() {
+	// Introduce a minor delay if there are any replications
+	// (sc.startReplicators() might rely on SG being fully started)
+	if len(sc.config.Replications) > 0 {
```
Comment (adamcfraser) [inline rest/server_context.go:94]: We're doing a length check here, and a nil check in startReplicators().  Looks like the nil check in startReplicators is incorrect - want to clean that up while you're here?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
One minor cleanup suggestion - otherwise looks good.

File: rest/server_context.go (hunk 1 above)
Comment (bbrks) [reply rest/server_context.go:94]: Done, thanks.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**